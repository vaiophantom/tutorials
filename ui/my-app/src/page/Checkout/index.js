import { InputGroup, Form, Button } from "react-bootstrap";
import Header from "../../components/Header";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import ProductCheckout from "../../components/ProductCheckout";
import { currencyFormat } from "../../utils/helper";
import { addToCart } from "../../utils/fetch";

const SUCCESS_MESS = "Your order was success !";

const Checkout = () => {
  const [listProducts, setListProducts] = useState([]);
  const navigate = useNavigate();
  const [total_payment, setTotal_payment] = useState(0);
  const [address, setAddress] = useState("");
  const [phoneNumber, setPhoneNumber] = useState("");

  useEffect(() => {
    let listCart = JSON.parse(localStorage.getItem("cart"));
    if (listCart) {
      listCart = listCart.map((item) => {
        return { checked: false, ...item };
      });
      setListProducts(listCart);
    }
  }, []);

  useEffect(() => {
    if (listProducts.length > 0) {
      const total = listProducts.reduce(
        (prev, current) => prev + current.price * current.quantity,
        0
      );
      setTotal_payment(total);
    }
  }, [listProducts]);

  const onOrder = async () => {
    const body = {
      productIds: listProducts.map((item) => item.id),
      totalPrice: total_payment,
      productQuantity: listProducts.length,
    };
    console.log(body);

    try {
      const res = await addToCart(body);
      if (res) {
        clearCart();
        navigate("/", { state: { message: SUCCESS_MESS } });
      }
    } catch (error) {
      console.log(error.message);
    }
  };
  const clearCart = () => {
    localStorage.removeItem("cart");
  };

  const onChangeAddress = (e) => {
    const { value } = e.target;
    setAddress(value);
  };

  const onChangePhone = (e) => {
    const { value } = e.target;
    setPhoneNumber(value);
  };
  const handleSearch = (type, value) => {};

  return (
    <div className="checkout-screen">
      <Header handleSearch={handleSearch} />
      <div className="checkout-body">
        <div style={{ flex: 1 }}>
          <Form.Label htmlFor="address">Address</Form.Label>
          <InputGroup hasValidation className="mb-3">
            <Form.Control
              required
              value={address}
              onChange={onChangeAddress}
              id="address"
              isInvalid={!address}
            />

            <Form.Control.Feedback type="invalid">
              Please fill a address.
            </Form.Control.Feedback>
          </InputGroup>

          <Form.Label htmlFor="phone-number">Phone Number</Form.Label>
          <InputGroup className="mb-3">
            <Form.Control
              isInvalid={!phoneNumber}
              value={phoneNumber}
              onChange={onChangePhone}
              type="tel"
              id="phone-number"
            />
            <Form.Control.Feedback type="invalid">
              Please fill a phone number.
            </Form.Control.Feedback>
          </InputGroup>

          <Form.Label htmlFor="">Payment Method</Form.Label>
          <InputGroup className="mb-3">
            <Form.Select aria-label="Default select example">
              <option value="1">Pay by cash</option>
              <option value="2" disabled>
                Pay online
              </option>
            </Form.Select>
          </InputGroup>
        </div>
        <div
          style={{
            flex: 1,
            paddingLeft: "40px",
            marginLeft: "40px",
            borderLeft: "1px solid rgb(237, 237, 237)",
          }}
        >
          <ul>
            {listProducts &&
              listProducts.map((item) => {
                return (
                  <li>
                    <ProductCheckout data={item} />
                  </li>
                );
              })}
          </ul>
        </div>
      </div>
      <div className="checkout-footer">
        <p style={{ fontWeight: "bold" }}>
          total price: {currencyFormat(total_payment)} $
        </p>
        <Button
          disabled={!address || !phoneNumber}
          type="submit"
          onClick={onOrder}
          variant="warning"
          style={{ margin: "20px" }}
        >
          Place to order
        </Button>
      </div>
    </div>
  );
};
export default Checkout;
