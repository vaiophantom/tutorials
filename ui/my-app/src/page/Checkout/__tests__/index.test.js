import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Checkout from '../';

it('renders Checkout', async () => {
 const {container} = render(<Checkout />, {wrapper: BrowserRouter});
 expect(container).toMatchSnapshot();
});
