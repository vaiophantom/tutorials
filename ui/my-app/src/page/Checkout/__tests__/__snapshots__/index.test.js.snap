// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Checkout 1`] = `
<div>
  <div
    class="checkout-screen"
  >
    <div
      class="header"
    >
      <img
        class="logo"
        height="100"
        src="logo.png"
        width="200"
      />
      <div
        class="search"
      >
        <div
          class="input-group"
          style="background-color: white; align-items: center;"
        >
          <button
            aria-expanded="false"
            class="dropdown-toggle btn btn-outline-primary"
            type="button"
          >
            name
          </button>
          <div
            aria-labelledby=""
            class="dropdown-menu"
            x-placement="bottom-start"
          >
            <a
              class="dropdown-item"
              data-rr-ui-dropdown-item=""
              href="#"
              role="button"
              tabindex="0"
            >
              name
            </a>
            <a
              class="dropdown-item"
              data-rr-ui-dropdown-item=""
              href="#"
              role="button"
              tabindex="0"
            >
              branch
            </a>
            <a
              class="dropdown-item"
              data-rr-ui-dropdown-item=""
              href="#"
              role="button"
              tabindex="0"
            >
              category
            </a>
            <a
              class="dropdown-item"
              data-rr-ui-dropdown-item=""
              href="#"
              role="button"
              tabindex="0"
            >
              color
            </a>
          </div>
          <input
            aria-describedby="basic-addon3"
            class="form-control"
            placeholder="Search products based on category name, product name, price, branch..."
            value=""
          />
          <img
            class=""
            height="30"
            src="search.jpeg"
            width="30"
          />
        </div>
      </div>
      <img
        class="cart"
        height="40"
        src="cart.png"
        width="40"
      />
    </div>
    <div
      class="checkout-body"
    >
      <div
        style="flex: 1;"
      >
        <label
          class="form-label"
          for="address"
        >
          Address
        </label>
        <div
          class="mb-3 input-group has-validation"
        >
          <input
            class="form-control is-invalid"
            id="address"
            required=""
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Please fill a address.
          </div>
        </div>
        <label
          class="form-label"
          for="phone-number"
        >
          Phone Number
        </label>
        <div
          class="mb-3 input-group"
        >
          <input
            class="form-control is-invalid"
            id="phone-number"
            type="tel"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Please fill a phone number.
          </div>
        </div>
        <label
          class="form-label"
        >
          Payment Method
        </label>
        <div
          class="mb-3 input-group"
        >
          <select
            aria-label="Default select example"
            class="form-select"
          >
            <option
              value="1"
            >
              Pay by cash
            </option>
            <option
              disabled=""
              value="2"
            >
              Pay online
            </option>
          </select>
        </div>
      </div>
      <div
        style="flex: 1; padding-left: 40px; margin-left: 40px; border-left: 1px solid rgb(237, 237, 237);"
      >
        <ul />
      </div>
    </div>
    <div
      class="checkout-footer"
    >
      <p
        style="font-weight: bold;"
      >
        total price: 
        0.00
         $
      </p>
      <button
        class="btn btn-warning"
        disabled=""
        style="margin: 20px;"
        type="submit"
      >
        Place to order
      </button>
    </div>
  </div>
</div>
`;
